### ISSUE TITLE
API simplification

### ISSUE THREAD

**Issue by fulmicoton** [tantivy-search/tantivy-py#9]
The current API sticks very closely to tantivy API.

Some of the choice of the current API are strongly specific Rust and do not necessarily make sense in python. Also, some choice of the current are actually not necessarily very smart even in the context of rust.

I suspect we could simplify our API a bit.
I explored differnet changes in https://github.com/tantivy-search/tantivy-py/pull/3, but I would like us to discuss them somewhere.

# Fields

Getting Field as objects obtained either at the creation of the `Schema` or by querying the `Schema` can be cumbersome. I suggest we let the user identify them as string all the way along.

Note this bit may change in rust's tantivy as well. 

Following the readme.

```python
builder = tantivy.SchemaBuilder()
title = builder.add_text_field("title", stored=True)
body = builder.add_text_field("body")
schema = builder.build()
```

We could remove saving the `title`, and `body` and even introduce a fluent interface. (note I do not know how to return self with PyO3. It might be impossible.)
```python
schema = tantivy.SchemaBuilder()
   .add_text_field("title", stored=True)
   .add_text_field("body")
   .build()
```

When referring to fields then, users could always directly pass in strings instead of the field objects.

# Documents

Conceptually, documents are just maps of field to a list of values.
We could remove the notion of document entirely and use dictionaries instead.

It might make onboarding much straightforward, and leave us with an arguably more appealing README.

That being said -as stated above-, tantivy allows more than one value per field.
Within tantivy json serializer, deserializer, the choice was made to allow for the single value format as an input, so that both
```json
{ "title": "The old man and the sea"}
```
and

```json
{ "title": ["The old man and the sea"]}
```

are valid documents.

On deserialization, we always deserialize as `{"title": ["The old man and the sea"]}`
This could lead to confusion within users, so there could be value in keeping a structured Document object. 

We could define a `__getitem__`  , `__len__` to make the user 's life easier while leaving him aware that the field can be multivalued.

# Reader

We could have the reader be part of the `Index`. 
The searcher would then be acquired directly from the `index`.

In a similar spirit, the query parsing could be directly methods of `Index`.
Also we could have a helper to perform a vanilla top-K search with or without count of documents directly from the index.

The README sample could become...

```
reader = index.searcher()
searcher = reader.searcher()

query = index.parse_query("sea whale", default_fields=[title, body])


search_results = searcher.search(query, nhits=10, count=True)
print(search_result.count)
(_score, doc_address) = search_results.hits()[0]
searched_doc = searcher.doc(doc_address)
assert searched_doc["title"] == ["The Old Man and the Sea"]

**Comment by poljar** [comment]
# Fields

The  fields suggestion exchanges a typed interface for fields with a stringly typed one. I'm on the fence if we want to do this, yes it will be simpler but slightly more error prone.

Even though I liked that we could store some important field values to be used later on in a search I understand that people are used to stringly typed interfaces in python so we might want to do this.

A fluent interface is possible, I don't know if we can get away without cloning here but it probably isn't too important since a schema is created once.

```rust
use pyo3::prelude::*;

#[pyclass]
#[derive(Clone, Copy)]
struct Fluent {
   num: i32,
}

#[pymethods]
impl Fluent {
    #[new]
    fn new(obj: &PyRawObject, num: i32) {
        obj.init({
            Fluent {
                num
            }
        });
    }

    fn add_one(&mut self) -> Self {
        self.num += 1;
        self.clone()
    }

    fn add_two(&mut self) -> Self {
        self.num += 2;
        self.clone()
    }

    #[getter(number)]
    fn num(&self) -> i32 {
        self.num
    }
}

#[pymodule]
fn fluent(py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<Fluent>()?;
    Ok(())
}
```

And the python would look like:
```python
x = fluent.Fluent()
assert x.number == 0
y = x.add_one().add_two()
assert y.number == 3
```

# Documents

The proposed documents API is used in a couple of [other](https://github.com/weixsong/elasticlunr.js#example) full text search libs. And it would make things simpler.

That being said we're again switching to a stringly typed API and I do prefer the idea to implement `__len__` and a dict like interface for the Document class.

I had some problems implementing some of the magic methods. I didn't investigate why it  wouldn't work with `cargo expand` and I forgot by now what it was about.

So we would need to check out if the dict like interface would work out.

# Reader

I like this idea, I implemented exactly this in a wrapper class. The wrapper class created the schema and opened the index. It also provided direct access to a searcher.

The query parsing idea is great as well,  as well as the helper function.

**Comment by poljar** [comment]
This has been addressed by #3 and #16. If someone has ideas for some other simplifications I suggest opening a separate issue for them.
